Cookbook/Proposals/ComposableReducers.md

```diff
@@ -0,0 +1,181 @@
+
+`Reducer` is a wrapper type around `reduce: (Action, State) -> State` function that conforms to `Monoid` (has "zero" and "+") to combine 2 reducers into one.
+
+By using this `append`ing capability, we can create more complex `Reducer` from splitted `SubReducer`s.
+
+### `Reducer` lifting from `SubState` / `SubAction`
+
+However, `SubReducer`s don't normally have the same type with the others, even with the `(Main)Reducer` type.
+
+For example,
+
+- Main screen: `MainReducer = Reducer<MainAction, MainState>`
+    - Component 1: `SubReducer1 = Reducer<Sub1Action, Sub1State>`
+    - Component 2: `SubReducer2 = Reducer<Sub2Action, Sub2State>`
+    - ...
+
+To convert `SubReducer1` and `SubReducer2` types into `MainReducer` (so that they can be combined using `<>`), we can use the following `lift` functions:
+
+```swift
+extension Reducer {
+    /// `Reducer<Action, SubState> -> `Reducer<Action, State>`
+    func lift<SuperState>(state lens: Lens<SuperState, State>) -> Reducer<Action, SuperState> {
+        return Reducer<Action, SuperState> { action, superState in
+            lens.set(superState, self.reduce(action, lens.get(superState)))
+        }
+    }
+
+    /// `Reducer<SubAction, State> -> `Reducer<Action, State>`
+    func lift<SuperAction>(action prism: Prism<SuperAction, Action>) -> Reducer<SuperAction, State> {
+        return Reducer<SuperAction, State> { superAction, state in
+            guard let action = prism.preview(superAction) else { return state }
+            return self.reduce(action, state)
+        }
+    }
+}
+```
+
+In short, to bring each small reducers to the same level and combine, we need `lift`.
+
+And to `lift`, we need `Lens` and `Prism`.
+
+## Example
+
+```swift
+// MARK: - Component 1 (isolated from Main & Component 2)
+//---------------------------------------------------------
+
+enum Sub1Action {
+    case increment
+    case decrement
+}
+
+struct Sub1State {
+    var count: Int = 0
+}
+
+let subReducer1 = Reducer<Sub1Action, Sub1State> { action, state in
+    switch action {
+    case .increment: return state.with { $0.count + 1 }
+    case .decrement: return state.with { $0.count - 1 }
+    }
+}
+
+// MARK: - Component 2 (isolated from Main & Component 1)
+//---------------------------------------------------------
+
+enum Sub2Action { ... }
+struct Sub2State { ... }
+let subReducer2: Reducer<Sub2Action, Sub2State> = ...
+
+// MARK: - Main
+//---------------------------------------------------------
+
+enum MainAction {
+    case sub1(Sub1Action)
+    case sub2(Sub2Action)
+
+    enum prism { // NOTE: Can codegen
+        static let sub1Action = Prism<MainAction, Sub1Action> = ...
+        static let sub2Action = Prism<MainAction, Sub2Action> = ...
+    }
+}
+
+struct MainState {
+    var sub1: Sub1State
+    var sub2: Sub2State
+    // var shared: ...  /* NOTE: Shared state can belong to here */
+
+    enum lens { // NOTE: Can codegen
+        static let sub1State = Prism<MainState, Sub1State> = ...
```

Comment:
shouldn't this be `Lens`? Also I thought KeyPaths are basically Lenses, isn't that so?

Reply:
Fixed in 5390a38 .